### PR TITLE
Handling Spans that are too Short & Increasing Timestamp Precision

### PR DESCRIPTION
**Pull request by ericthewry**

### CODE REVIEW

**Comment by ericthewry** [comment]
@ahorn

**Comment by ahorn** [comment]
Thanks for sending this pull request. Can you please rebase as Aaron described in the meeting. This simplifies the code review.

**Comment by ericthewry** [comment]
@ahorn Done; this should simplify things!

**Comment by ahorn** [comment]
Thank you. There is also a commit squash feature [1].

[1] https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

**Comment by ericthewry** [comment]
Would you like this as a single commit?

**Comment by ahorn** [comment]
If you think one captures the discrete changesets, then yes. Commits that reflect work in progress or stretch work are confusing.

**Comment by ericthewry** [comment]
> If you think one captures the discrete changesets, then yes. Commits that reflect work in progress or stretch work are confusing.

Gotcha!

> Don't create setters and getters. Use const qualifiers instead.

My goal here is to enforce the [condition](https://github.com/ahorn/nopticon/pull/5/commits/f9105368dae47c95c88b9612dc72b477e5eea2f7#diff-dbdc39af94b24c75be724e2a2f340fc7R30) `d <= m_span` on insert. The most consistent way to do this is to encapsulate the check with the assignment action, e.g. via a setter method. I do not think that there is a way to perform this check using const qualifiers (unless we Ctrl-f for every occurrence of the assignment), but please correct me if I am wrong.

The Getter-Setter approach fixed a bug where duration was much greater than m_span resulting in ranks greater than 1.0.  Enforcing this getter-setter pattern eliminated the bug.

I found [a compromise](https://stackoverflow.com/a/14540455) that basically consists of creating a new type for duration (that knows about its containing span size). This new type would have an overwritten assign method, which where we could insert the check that d <= m_span. This way, duration could just be public! What do you think?

**Comment by ahorn** [comment]
I have only glanced at the changes.

How is this check in the setter different from the assertion that duration <= span [1]? What was the reason for making the tail member field public? Please preserve white space to avoid confusing the patch. If you want debug messages, please guard them by an appropriate ifdefs through a DEBUG macro or use assertions plus the &&-idiom. I have not had a chance yet to review the meat of the change that concerns the removing of the std::exit call.

[1] https://github.com/ahorn/nopticon/pull/5/commits/f9105368dae47c95c88b9612dc72b477e5eea2f7#diff-5d28fd3bfaf703597d51e3134c03ee0dR163

File: test/analysis_test.cc [hunk 1]
```diff
@@ -422,12 +436,25 @@ static void test_intersection_of_timestamps() {
 }
 
 void run_analysis_test() {
```
Comment (ahorn) [inline test/analysis_test.cc]: I appreciate the effort to put in the text. There is a more idiomatic way of doing it by declaring a test function through a macro and that macro will be automatically print the name of the function when it is executed. Let's hold off these manually inserted output strings and I can change the tests to use the more idiomatic way.

Comment (ericthewry) [reply test/analysis_test.cc]: :+1:

**Comment by ericthewry** [comment]
> How is this check in the setter different from the assertion that duration <= span [1]? 
> [1] [f910536#diff-5d28fd3bfaf703597d51e3134c03ee0dR163](https://github.com/ahorn/nopticon/commit/f9105368dae47c95c88b9612dc72b477e5eea2f7#diff-5d28fd3bfaf703597d51e3134c03ee0dR163)

:+1: You're right. They're not different. Those are leftover asserts that I put in to try and find the rank > 1.0 bug. These asserts passed even while I was observing ranks greater than 1 in the end-to-end output, so they're quite useless. I'll remove them.

> What was the reason for making the tail member field public? 

`history_t` is no longer a friend class, because I wanted to prevent it from arbitrarily changing `slice_t::duration`.

> Please preserve white space to avoid confusing the patch.

👍 I tend to like more whitespace than you do, so I often add it to improve readability. That being said, sometimes I also leave extra tabs hanging around.  I'll be sure to remove these in the future. 

>  If you want debug messages, please guard them by an appropriate ifdefs through a DEBUG macro or use assertions plus the &&-idiom. 

I'm not familiar with the &&-idiom. I'll use a DEBUG macro.

**Pull request closed without merging**